Mathlib/NumberTheory/Cyclotomic/PrimitiveRoots.lean

```diff
@@ -189,6 +189,101 @@ theorem finrank (hirr : Irreducible (cyclotomic n K)) : finrank K L = (n : ℕ).
     (zeta_spec n K L).minpoly_eq_cyclotomic_of_irreducible hirr, natDegree_cyclotomic]
 #align is_cyclotomic_extension.finrank IsCyclotomicExtension.finrank
 
+variable {L} in
+/-- If `L` contains both a primitive `p`-th root of unity and `q`-th root of unity, and
+`Irreducible (cyclotomic (lcm p q) K)` (in particular for `K = ℚ`), then the `finrank K L` is at
+least `(lcm p q).totient`. -/
+theorem _root_.IsPrimitiveRoot.lcm_totient_le_finrank [FiniteDimensional K L] {p q : ℕ} {x y : L}
+    (hx : IsPrimitiveRoot x p) (hy : IsPrimitiveRoot y q)
+    (hirr : Irreducible (cyclotomic (Nat.lcm p q) K)) :
+    (Nat.lcm p q).totient ≤ FiniteDimensional.finrank K L := by
```

Comment:
This looks slightly strange to me: existence of a primitive p'th root and a primitive q'th root is the same as existence of a primitive `Nat.lcm p q`th root, so this theorem feels like it's the composite of two theorems, one saying that and the other saying that if there exists a primitive x'th root then (change all `Nat.lcm p q` to `x`). And why do you demand the primitive roots rather than just asserting their existence? They don't play any role in the other hypotheses or conclusion.

Reply:
Would it work to just pass around `\exists x, IsPrimitiveRoot x n`?

Reply:
Splitting the proof is surely a good idea, I am doing it.

Concerning the existence vs giving the witness, I guess this goes back to things like [IsPrimitiveRoot.card_rootsOfUnity'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/RootsOfUnity/Basic.html#IsPrimitiveRoot.card_rootsOfUnity'): if we refactor that file to use `∃ ζ, ...` then we will probably spot all places where we need to change it. I indeed think it is a good idea to use existence rather than the explicit witness (did I really ask to do the opposite? Sorry for that), but probably in another PR.

Reply:
Sounds good to me.

Reply:
I've golfed the proof using the explicit root of unity of order the `lcm` provided in #11235. The proof is indeed much shorter now, thanks!

Reply:
I know I am late to the game, but several of these results seem to hold beyond the case of cyclotomic extensions of fields (not all of them, I agree, but the above inequality holds more generally). I have tried to generalize the proof to e.g. integral domains, and everything works but  at one point (the devil in the details...) where things become a bit messy since one has to trace back all places where the field assumption was not really needed. What is the general philosophy?

Reply:
Note that `IsPrimitiveRoot.lcm_totient_le_finrank` (the inequality) does not assume the extension is cyclotomic. It is in this file because it uses cycolotomic extension in the proof.

Concerning the domain/field assumption, it seems to me that it is `IsCyclotomicExtension.finrank` that requires a field, and it can be trivial to generalize. I don't have much time now, but I can have a look tomorrow.

Reply:
Mmm, already `IsPrimitiveRoot.powerBasis` wants a field. We may use [Algebra.adjoin.powerBasis'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Minpoly/IsIntegrallyClosed.html#Algebra.adjoin.powerBasis'), but it wants a noetherian ring. Maybe we can try the generalization in another PR?

Reply:
Here the goal is purely number theoretic: for example if `p=3`, then any unit in `ℤ[ζ₃]` is a root of unity (for example using Dirichlet... ok, this is maybe like using an atomic bomb, but still) and then we immediately (at least it is mathematically trivial to) get the list of the 6 units. And then we can prove kummer's lemma by hand.